KAFKA-7250: switch scala transform to TransformSupplier

vvcephei opened the pull request:
#5468 introduced a breaking API change that was actually avoidable. This PR re-introduces the old API as deprecated and alters the API introduced by #5468 to be consistent with the other methods

* also, fixed misc syntax problems

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
MINOR: fix-kstreams-interface

vvcephei commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala:
```diff
@@ -0,0 +1,149 @@
+    override def close(): Unit = ???
+  }
+
+  private class MockJStream extends JStream[String, Long] {
```
I tried to use EasyMock, but the capture wouldn't work; I think because of the varargs parameter.

vvcephei commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:
```diff
@@ -207,17 +204,20 @@ class TopologyTest extends JUnitSuite {
-            }
-          )
+              }
+          })
```
I also compiled with scala 2.12 to verify that the old code works via SAM conversion. I don't think we can explicitly test with it in this project, though, until we update to 2.12.

vvcephei commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -295,9 +297,42 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * @see `org.apache.kafka.streams.kstream.KStream#transform`
    */
-  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, KeyValue[K1, V1]],
+  def transform[K1, V1](transformerSupplier: TransformerSupplier[K, V, KeyValue[K1, V1]],
```
Question: should we reform the process method to also expect a `ProcessorSupplier`? It shouldn't break anyone on 2.12+ due to SAM conversion.

mjsax replied:
I don't think we should change `process` because it would require a KIP. We should rather fix the bug only. Maybe be can create a Jira for it though.

vvcephei commented:
@guozhangwang @mjsax : can you take a look at this? thanks.

@mdziemianko : We had some tests fail in another project due to the API breakage. I had an idea for how to preserve the pre-existing API while also fixing the bug you fixed. Do you mind sharing your thoughts on this PR?

guozhangwang commented on build.gradle:
```diff
@@ -709,7 +709,6 @@ project(':core') {
                                'genAdminClientConfigDocs', 'genProducerConfigDocs', 'genConsumerConfigDocs',
                                'genKafkaConfigDocs', 'genTopicConfigDocs',
                                ':connect:runtime:genConnectConfigDocs', ':connect:runtime:genConnectTransformationDocs',
-                               ':connect:runtime:genSinkConnectorConfigDocs', ':connect:runtime:genSourceConnectorConfigDocs',
```
Are these change intentional? Or maybe you need to rebase from latest trunk..

vvcephei replied:
woah! it looks like I need to rebase.

guozhangwang commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -176,15 +176,17 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    */
-  def branch(predicates: (K, V) => Boolean*): Array[KStream[K, V]] =
+  //noinspection ScalaUnnecessaryParentheses
+  def branch(predicates: ((K, V) => Boolean)*): Array[KStream[K, V]] = {
```
Why this change?

vvcephei replied:
The type checker in IDEA was spazzing out and treating `predicates`'s type as `Seq[(K,V) => Any]`.
This was the only way I could get it to shut up.

I actually couldn't compile this class in IDEA without this change.

Funny enough, it fails without the parens and then complains that the parens are unnecessary once you add them.

Can you check in your IDEA that you get no reported error and that you can run the tests I added from IDEA? If so, maybe I have the ide misconfigured.

guozhangwang commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -211,7 +213,7 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * }}}
    *
    * @param topic the topic name
-   * @param (implicit) produced the instance of Produced that gives the serdes and `StreamPartitioner`
```
Why removing `implicit`?

vvcephei replied:
It's a scaladoc syntax error.

vvcephei replied:
We could add it after `produced`, but AFAICT, the param name must immediately follow `@param`

vvcephei replied:
unless my IDE is also wrong about this...

mjsax replied:
You can build the scala docs and double check if in doubt -- if is works same a JavaDocs, I think John is correct.

vvcephei replied:
Confirmed: when I generate the scaladoc from trunk, it just skips this parameter and only documents `topic`, but when I generate it from this branch, it documents both `topic` and `produced`.

guozhangwang commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -295,9 +297,42 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
+        override def get(): Transformer[K, V, KeyValue[K1, V1]] =
+          new Transformer[K, V, KeyValue[K1, V1]] {
+            override def transform(key: K, value: V): KeyValue[K1, V1] =
+              transformer.transform(key, value) match {
```
We should add in the document that users should not use a transformer instance as the parameter.

vvcephei replied:
done.

vvcephei replied:
What do you think about just throwing an explanatory exception instead?

guozhangwang commented on streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  // demonstrating the condition under which the deprecated method yields incorrect behavior
+  @Test def deprecatedTransformShouldDoesntCreateATrueSupplierWhenPassedAnInstance(): Unit = {
```
nit: `..ShouldNot..`

vvcephei replied:
haha, fair enough!

guozhangwang reviewed (commented):
The proposed change lgtm. As for the deprecated API, if people feel that directly removing it is too harsh we can leave it, though personally I'd prefer just remove it with the new API that would not potentially leave buggy holes for users.

Have you validate this change with splotlessCheck?

vvcephei committed:
cr comments

vvcephei committed:
spotless

vvcephei commented:
@guozhangwang what do you think about this:

> Question: should we reform the process method to also expect a ProcessorSupplier? It shouldn't break anyone on 2.12+ due to SAM conversion.

vvcephei committed:
optimize imports

mdziemianko commented:
LGTM too.

I was hoping to have time to go over the whole DSL and make it more consistent as each function looks different:) I chose the generator function because it is similar to how map, join etc work. The transformValue are the only ones that accept suppliers as far as I can tell from a brief look over the code.

I am of an opinion that this DSL looks a bit like PHP standard library... each signature is different. It goes not only for object vs generator vs supplier, but also how the serdes are accepted (especially in KTable), for example:
def filter(predicate: (K, V) => Boolean, materialized: Materialized[K, V, ByteArrayKeyValueStore]): KTable[K, V]
versus:
def groupBy[KR, VR](selector: (K, V) => (KR, VR))(implicit serialized: Serialized[KR, VR]): KTable[K, VR] 
or my favourite:
def join[VO, VR](other: KTable[K, VO])(joiner: (V, VO) => VR, materialized: Materialized[K, VR, ByteArrayKeyValueStore]): KTable[K, VR] 

The point being - pick one scheme, I am willing to spend some time making sure that everything adhere to this scheme...

I also think a buggy discontinued method should be just deleted rather than deprecated especially in a DSL that just come out and not many people are using it yet.

vvcephei committed:
license

guozhangwang commented:
@mdziemianko The reason of the inconsistency in KTable function signatures is for implicit conversions: we want to ease users of specifying those parameters that can be inferred by Scala anyways. And for different functions the parameters that can be implicit are a bit different.

@vvcephei For `process()`, if we change the signature would that break for users on Scala 2.11?

vvcephei commented:
@mdziemianko Thanks for the review.

I think we would welcome a PR to standardize the scala DSL. Please do, if you find the time!

On the deprecate-vs-delete, I think I tend to err on the side of deprecating in general, but this may be a special case. Special factors might be:
* the extreme youth of this API
* the fact that the interface itself is buggy

Usually, we can deprecate an unfortunate method and put in some kind of correct workaround implementation, but in this case, there's nothing we can do if you pre-instantiate a transformer and pass it in. We can't even detect it.

The by-reference argument makes it slightly better, since it *might* really act like a factory, but there's no guarantee.

Maybe it is better to just delete it in this case... 

Or we could keep the method but throw an exception in the body explaining the problem. I just imagine bumping up to a new bugfix version of 2.0.x is going to fail to compile, and we won't get a chance to explain why such an extreme measure was necessary. Of course, it'll be in the release notes, but I'd feel better if it was "in the loop" for the dev.

vvcephei commented:
@guozhangwang yes, it would, but we could add an implicit conversion to ease the transition.

Also note that process is now the only method in the whole API that takes `() => X` instead of `XSupplier`.

EDIT: I take that back; according to my local tests, the two methods can exist side-by-side, so we can keep the `()=>X` version and add a `XSupplier` version. So no one would be broken.

vvcephei committed:
add process variant taking a supplier

vvcephei commented:
@guozhangwang Since it looks fine to add the process variant taking a supplier, I've gone ahead and added it.

mdziemianko commented:
@guozhangwang actually, I just converted quite a bit chunk of code from lightbend DSL wrapper and I have a feeling that compiler was able to infer more types and convert/wrap functions to suppliers on its own.  I had to add quite a few type annotations, mostly (i think) around areas where FunctionConversions kick in but I was not paying too much attention given the amount of code I had to convert on short notice...

Is there any good reason why certain functions such as aggregate or join have the parameters list split? not talking about implicits but about regular parameters split into multiple params list. Are you expecting a lot of partial applications?

mjsax commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -295,9 +296,45 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
+   * @see [[org.apache.kafka.streams.kstream.KStream#transform]]
+   */
+  @deprecated("Use transform(TransformerSupplier,String) instead.", "2.0")
+  def transform[K1, V1](transformer: => Transformer[K, V, (K1, V1)], stateStoreNames: String*): KStream[K1, V1] = {
```
I don't think we need to keep a deprecated method -- the old method allows to compile code, but it will fail during runtime. Thus, nobody can actually use the old method and thus there is no reason to keep it.

mjsax commented:
@John: btw: the PR title should reflect the Jira. It's not MINOR. KAFKA-7250 -- also, please reference this PR in a comment on the JIRA. Thx.

vvcephei committed:
revert process and drop broken transform

vvcephei commented:
@mjsax Thanks, I've renamed the PR, and I'll add a comment to the Jira. I have also created https://issues.apache.org/jira/browse/KAFKA-7267 with my thoughts about the process method.

vvcephei commented:
@mdziemianko @guozhangwang @mjsax 

I've updated this PR to drop the broken transform method (just as #5468 did). I thought that we could keep it using a by-reference parameter, but that API still couldn't be guaranteed to be correct.

@mjsax also pointed out that the existing transform method has been unusable since it was introduced, which is proof that no one is using it and therefore, it's ok to break the API to fix the bug.

Nevertheless, I've decided to keep the PR just to clean up a few things in the Scala API and also to use a `TransformerSupplier` instead of a `() => Transformer`. This is really a 6-one-way-half-dozen-the-other kind of thing, but here's my thought:
* Both `KTable#transform` and `KStream#transformValues` take suppliers. The only other method that takes a function is `KStream#process`. Given that we have to change transform to something, I read this as a 2:1 vote in favor of supplier.
* users starting from scala 2.12 can pass a function to a method expecting a supplier, and SAM conversion will make it work, without even needing an implicit conversion. If we expect a function, though, there's no automatic conversion from supplier.

Note also that I previously had a test, which I've removed, since there's nothing left in this PR for it to test.

vvcephei commented on streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:
```diff
@@ -31,8 +31,8 @@ import scala.collection.JavaConverters._
  * Wraps the Java class [[org.apache.kafka.streams.kstream.KStream]] and delegates method calls to the underlying Java object.
  *
- * @param [K] Type of keys
- * @param [V] Type of values
```
This was also invalid scaladoc and caused the type parameters not to be documented at all.

guozhangwang commented:
LGTM. Waiting for jenkins to pass.

guozhangwang commented:
Cherry-picked to 2.0 as well.
